### PR TITLE
[intesis] Improve session handling

### PR DESCRIPTION
**Pull request by fd0cwp**
I have changed the login process so that the binding now only logs in at the beginning and logs out at the end. If a request fails due to a missing login or an invalid session ID, a new login attempt is made.

This ensures that
- Several requests can be sent in direct succession (if I switch the air conditioning from cooling to heating, I also want to change the slats at the same time)
- Less traffic is generated as there is not a constant logout and login again
- Ideally, the login credentials should only be visible once at startup

I am unsure whether I have used the correct levels for the logging.

The updated binding has been running without any problems for a few weeks now in my environment. First under 4.1.0, and since a few days now under 4.1.1.

Signed-off-by: Christoph <fd0cwp@gmx.de>

### CODE REVIEW

File: bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java [hunk 1]
```diff
@@ -208,39 +216,43 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+            m_sessionId = "";
         }
+        return m_sessionId;
     }
 
     public @Nullable String logout(String sessionId) {
         String contentString = "{\"command\":\"logout\",\"data\":{\"sessionID\":\"" + sessionId + "\"}}";
-        return api.postRequest(config.ipAddress, contentString);
+        String response = api.postRequest(config.ipAddress, contentString);
+        return response;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:255]:
```suggestion
        return api.postRequest(config.ipAddress, contentString);
```

File: bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java [hunk 2]
```diff
@@ -309,18 +321,34 @@ public void addChannel(String channelId, String itemType, @Nullable final Collec
...
-                    boolean success = resp.success;
-                    if (success) {
+                    if (resp.success) {
                         handler.accept(resp);
                     } else {
                         updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                                 "Request unsuccessful");
+                        ResponseError respError = gson.fromJson(resp.error, ResponseError.class);
+                        if (respError != null) {
+                            logger.error("postRequest failed - respErrorCode: {} / respErrorMessage: {} / retry {}",
```
Comment (lsiepel) [inline bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:341]: This should be changed to warn. Can't fully asses in the limited time i have, but if this is more then a transient error,  this error should bubble up and eventually set the Thing state. Hope this makes sense.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Left some comments, the build error has to be fixed (spotless), please also check if there are compile warnings.

Edit: Documentation about logging is here: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java [hunk 3]
```diff
@@ -86,6 +87,8 @@ public class IntesisHomeHandler extends BaseThingHandler {
 
     private IntesisHomeConfiguration config = new IntesisHomeConfiguration();
 
+    private String m_sessionId = "";
```
Comment (lsiepel) [inline bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:90]:
```suggestion
    private String sessionId  = "";
```

SAT Error: Name 'm_sessionId' must match pattern '^[a-z][a-zA-Z0-9]*$'.

Comment (fd0cwp) [reply bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java]: Hi @lsiepel,
Thank you for the fasst review!
What means "SAT Error"?
I wanted to make clear what is the global variable and what is a local one. But if you think this is not reasonable, it is OK for me. So I check in an updated version.

Comment (lsiepel) [reply bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java]: Details can be found here: https://www.openhab.org/docs/developer/guidelines.html#static-code-analysis
The report is available in your local build, but also on this PR you can click details next to the ci build check, click summary and the download the sat-summary-report : https://github.com/openhab/openhab-addons/actions/runs/8154201322?pr=16476

File: bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java [hunk 4]
```diff
@@ -109,6 +112,9 @@ public void initialize() {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Password not set");
             return;
         } else {
+            logger.trace("trying to log in - current session ID: {}", sessionId);
+            login();
```
Comment (jlaur) [inline bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:116]: `initialize()` should return fast and not do any blocking network calls. This is the reason for background initialization below.

File: bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java [hunk 5]
```diff
@@ -129,6 +135,8 @@ public void dispose() {
             refreshJob.cancel(true);
             this.refreshJob = null;
         }
+
+        logout(sessionId);
```
Comment (jlaur) [inline bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:139]: Same here.

**Comment by lsiepel** [comment]
@fd0cwp can you come up with a folow up PR as i missed these comment jlaur mentioned.

**Comment by fd0cwp** [comment]
I am not sure if this is reasonable. 

What is the purpose of the initialize function? I would say that the binding shall be set up so that it is functional. In my view, this also includes the login. 
initialize should only be run when openHAB is started. Does the two seconds for the login matter? Or am I missing something?
From my point of view, the same applies to dispose.

I could remove the login from the initialize. This would result in the first call failing, then the login is made and the call is tried again.
I could also remove the logout from dispose. But then openHAB no longer shuts down cleanly.

@lsiepel, @jlaur: What do you think?

**Comment by jlaur** [comment]
> @lsiepel, @jlaur: What do you think?

I already stated what I think. 🙂 Please have a look at https://www.openhab.org/docs/developer/bindings/#startup:
> Furthermore, the framework expects `initialize()` to be non-blocking and to return quickly. For longer running initializations, the implementation has to take care of scheduling a separate job which must guarantee to set the status eventually.

So this means you need to move the `login();` into the `scheduler.submit` block just below, as mentioned. Similarly, you can also schedule the logout - like it was previously also done here:
https://github.com/openhab/openhab-addons/commit/20f306f4856e9e348421d92361d560997336f6ca#diff-94852e57bc82cccaead7afd2af8297593bd11fd5cee89fa0957a41a57677494eL176-L182

**Pull request merged**